MaxText/goodput.py

```diff
@@ -0,0 +1,127 @@
+"""
+Goodput Prototype (TODO: Move to packaged library)
+"""
+
+import os
+import max_logging
+import time
+import jax
+import datetime
+import json
+from google.cloud import storage
+import logging
+logger = logging.getLogger(__name__)
+step_time_start_dict = {}
+step_time_filename = f"goodput_step_times.txt"
+job_runtime_filename = f"goodput_job_runtime.txt"
+
+def should_log_goodput(config):
+  """Determine whether or not to enable Goodput logging"""
+  if jax.process_index == 0 and config.enable_goodput_metrics:
+      return True
+  return False
+
+def get_goodput_dir(config):
+  """Derive the goodput directory where logs will be stored"""
+  goodput_dir = os.path.join(config.base_output_directory, config.run_name, "goodput", "")
+  return goodput_dir
+
+def get_gcs_bucket_and_blob_name(destination_gcs_name):
+  """Derive the GCS blob and bucket name"""
+  path_parts = destination_gcs_name.replace("gs://", "").split("/")
+  bucket = path_parts.pop(0)
+  blob = "/".join(path_parts)
+  return bucket, blob
+
+def upload_file_to_gcs(destination_gcs_name, source_file_name):
+  """Upload a file to a GCS location"""
+  try:
+    bucket_name, blob_name = get_gcs_bucket_and_blob_name(destination_gcs_name)
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(blob_name)
+    blob.upload_from_filename(source_file_name)
+  except Exception as e:  # pylint: disable=broad-exception-caught
+    logger.error("Error in uploading file to GCS location.", e)
+
+def download_file_from_gcs(source_gcs_name, destination_file_name):
+   """Download a file from a GCS location"""
+   try:
+      bucket_name, blob_name = get_gcs_bucket_and_blob_name(source_gcs_name)
+      storage_client = storage.Client()
+      bucket = storage_client.bucket(bucket_name)
+      blob = bucket.blob(blob_name)
+      blob.download_to_filename(destination_file_name)
```

Comment:
BTW this type of logic is unwelcome on our control thread! We need to be writing asynchronously to something that gracefully handles appends.

(GCS really doesn't like appends. And this code seems to be synchronous on our control thread.)